cpp/src/jni/dataset/jni_util.h

```diff
@@ -0,0 +1,143 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include "arrow/array.h"
+#include "arrow/io/api.h"
+#include "arrow/ipc/api.h"
+#include "arrow/memory_pool.h"
+#include "arrow/result.h"
+#include "arrow/type.h"
+
+#include <jni.h>
+
+namespace arrow {
+namespace dataset {
+namespace jni {
+
+class JniPendingException : public std::runtime_error {
```

Comment:
does this need to be in the header?  Please add docs?

Reply:
Moved these error handling tools back to jni_wrapper.cc.

I was thinking of preparing some common utilities for both `jni/dataset` `jni/orc` (or maybe `gandiva/jni`). But we can actually do this later in some other tickets. The utilities also can not work well without the macros in `jni_wrapper.cc`.

Added some docs to other interfaces in this header.

Reply:
yeah, I think this is a good thought and it would be great to see some refactoring here to share common code.  Thank you for taking the time to address all the concerns around this.